File: 시즌 11/중/kimryungyo/other_languages/day_7.java
// 자바가 재귀에서 느리다는 얘기를 듣긴 했지만 진짜 심각할 정도로 느리구나..

import java.util.Scanner;

public class Main {

    static void dfs(int N, int M, int[] sequence, boolean[] visited, int depth) {
        if (depth == M) {
            for (int num : sequence) {
                System.out.print(num + " ");
            }
            System.out.println();
            return;
        }

        for (int i = 1; i <= N; i++) {
            if (!visited[i]) {
                sequence[depth] = i;
                visited[i] = true;
                dfs(N, M, sequence, visited, depth + 1);
                visited[i] = false;
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int M = sc.nextInt();
        boolean[] visited = new boolean[N + 1];
        dfs(N, M, new int[M], visited, 0);
    }

}

File: 시즌 11/중/kimryungyo/other_languages/day_7.c
#include <stdio.h>

int N, M;
int visited[9];

void dfs(int sequence[], int length) {
    if (length == M) {
        for (int i = 0; i < M; i++) {
            printf("%d ", sequence[i]);
        }
        printf("\n");
        return;
    }

    for (int num = 1; num <= N; num++) {
        if (!visited[num]) {
            sequence[length] = num;
            visited[num] = 1;
            dfs(sequence, length + 1);
            visited[num] = 0;
        }
    }
}

int main() {
    scanf("%d %d", &N, &M);

    int sequence[10];
    dfs(sequence, 0);

    return 0;
}